qiskit_experiments/calibration_management/basis_gate_library.py

```diff
@@ -27,49 +29,57 @@
 from qiskit_experiments.exceptions import CalibrationError
 
 
-class BasisGateLibrary(ABC):
+class BasisGateLibrary(ABC, Mapping):
     """A base class for libraries of basis gates to make it easier to setup Calibrations."""
 
     # Location where default parameter values are stored. These may be updated at construction.
     __default_values__ = dict()
 
-    # The basis gates that this library generates. This is a dict with the gate name as key
-    # and the number of qubits that gate acts on as values.
-    __supported_gates__ = None
-
     def __init__(
-        self, basis_gates: Optional[List[str]] = None, default_values: Optional[Dict] = None
+        self,
+        basis_gates: Optional[List[str]] = None,
+        default_values: Optional[Dict] = None,
+        **extra_kwargs,
     ):
         """Setup the library.
 
         Args:
             basis_gates: The basis gates to generate.
             default_values: A dictionary to override library default parameter values.
+            extra_kwargs: Extra key-word arguments of the subclasses that are saved to be able
+                to reconstruct the library using the :meth:`__init__` method.
 
         Raises:
             CalibrationError: If on of the given basis gates is not supported by the library.
         """
-        self._schedules = dict()
-
         # Update the default values.
-        self._default_values = dict(self.__default_values__)
+        self._extra_kwargs = extra_kwargs
+        self._default_values = self.__default_values__.copy()
         if default_values is not None:
             self._default_values.update(default_values)
 
         if basis_gates is None:
-            self._basis_gates = self.__supported_gates__
+            basis_gates = list(self.__supported_gates__)
 
-        else:
-            self._basis_gates = dict()
+        self._schedules = dict()
+        for gate in basis_gates:
+            if gate not in self.__supported_gates__:
+                raise CalibrationError(
+                    f"Gate {gate} is not supported by {self.__class__.__name__}. "
+                    f"Supported gates are: {self.__supported_gates__}."
+                )
+
+        # Populate self._schedules based on the init arguments.
+        self._build_schedules(set(basis_gates))
```

Comment:
Could we change this to:

```suggestion
        self._schedules = self._build_schedules(set(basis_gates))
```

This would allow for `_build_schdules` to document the types of the keys and values of `self._schedules` more clearly. At the moment, it is a little difficult to figure out what `self._schedules` should be.

Reply:
Good point. Done here: aaf37f2